compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/TypeReference.kt

```diff
@@ -52,12 +70,38 @@ public sealed class TypeReference {
         ?.toClassReference(module)
     }
 
+    override val typeName: TypeName? by lazy(NONE) {
+      try {
+        type.requireTypeName(module)
+      } catch (e: AnvilCompilationException) {
+        // The goal is to inline the function above and then stop throwing an exception.
+        null
+      }
+    }
+
     override fun resolveGenericTypeOrNull(implementingClass: ClassReference.Psi): ClassReference? {
       classReference?.let { return it }
 
       val typeReference = implementingClass.resolveTypeReference(type) ?: type
       return typeReference.fqNameOrNull(module)?.toClassReference(module)
     }
+
+    override fun resolveGenericTypeNameOrNull(implementingClass: ClassReference.Psi): TypeName? {
+      val typeReference = implementingClass.resolveTypeReference(type) ?: type
+      return try {
+        typeReference.requireTypeName(module)
+          .let { typeName ->
+            // This is a special case when mixing parsing between descriptors and PSI.  Descriptors
+            // will always represent lambda types as kotlin.Function* types, so lambda arguments
+            // must be converted or their signatures won't match.
+            // see https://github.com/square/anvil/issues/400
+            (typeName as? LambdaTypeName)?.asFunctionType() ?: typeName
+          }
+      } catch (e: AnvilCompilationException) {
+        // The goal is to inline the function above and then stop throwing an exception.
```

Comment:
Same here for adding TODO

Reply:
I removed one of these methods already in the next PR. I'll add a TODO to the remaining method in the other stacked PR.